xgcm/test/test_derivatives.py

```diff
@@ -8,23 +8,43 @@ def test_derivative_uniform_grid():
     # this is a uniform grid
     # a non-uniform grid would provide a more rigorous test
```

Comment:
I have refactored a similar one in `tests/datasets/datasets_metric`, could you see if that works for this test? That way we can change that to a nonuniform spacing and have it update in all tests.

Reply:
Nice! I was able to use that, see the latest push for a C grid derivative tests. A couple of comments on this: 
1. Could easily also add a B grid test if you think it's necessary
2. For some reason, the `.equals( )` always returns `False` for me, even when the absolute value of the difference between `expected` and the test derivative is zero. This may be my version of pytest, or something, but I'd be interested if you understand why or see something I'm missing. 
3. There is no Z axis derivative for the velocity points. I even tried to add a `dz_e` living at `(xu, yt, zw)`, and similarly for `dz_y`, but xgcm couldn't figure out the metric to use. Is this because we would need another axis, like 'ZE' or something in addition to 'Z'? I'm thinking this is because there is no defined xgcm axis for these points on the grid.

Reply:
1) I think a B-grid test would be great. Could you add that? It seems to me that `test_derivative_uniform_grid():` is now pretty similar to the other test? Perhaps you can delete that test and include both c and b grid test in a `test_derivative` function? @rabernat what is your style opinion here?
2) Weird. Could it be that the coordinates/attributes are somehow different? Does [this](https://github.com/xgcm/xgcm/blob/cac66afeb2426546b142aef07b7752bd02ea0ebb/xgcm/test/test_metrics.py#L34) make it pass?
3) Could you explain a bit more (or share a notebook or code) how you added those metrics? I am surprised that it didnt handle it, and we should definitely find out why.

Reply:
[This](https://github.com/xgcm/xgcm/pull/130/files/415e89f0a9c69ae39666910de14b26f4985f11ae#r303041847) seems to answer what Ryan thinks about the style of tests. So please disregard 
>Perhaps you can delete that test and include both c and b grid test in a test_derivative function?

Reply:
I couldn't see the link to what you meant about what Ryan's test style preference, but I'm assuming you meant to not delete the old `test_derivative_uniform_grid()` function. Let me know if you meant something else. 

1. B grid added with latest commit. I could have written both B and C grid in a more nested for loop structure, but I think this is nicer because things are more explicit. Let me know if you agree with that.

2. Nice catch! Yeah that did the trick and I made that change. 

3. Coming soon...

Reply:
Re 3: Turns out this is no problem! I just had to draw myself a picture of a finite volume grid to do it right. I ended up changing how the `dz` elements are created, I made them now 1D vectors just like they are in e.g. MITgcm. Let me know if this breaks anything. 

I made that change after doing things more complicated the first go around. I rolled all of that back, but here's a description of one of the intermediate commits just in case you're curious...

b55986e added the `dz` elements which live on each of horizontal velocity points, and while doing that I changed the way all the `dz` elements are referenced in order to be as clear as possible since I think it got a bit confusing. I mentioned it in the comments, but it works like: dz_<1>_<2> where <1> refers to location in horizontal, <2> location in vertical. So the old `dz_t` -> `dz_t_t` and `dz_w` -> `dz_t_w`. After realizing the time dimension was in the `dz` elements, I realized we can just make them 1D vectors, so the additions here don't seem necessary and just make things confusing.

Reply:
Ok it looks like changing `dz` to 1D vectors breaks `test_metrics.test_get_metric()` in the case where 

`axes=('X', 'Y', 'Z'), data_var= 'u' , drop_vars=None, metric_expected_list=['volume_t'], expected_error=KeyError` 

However, in that test `grid.get_metric()` now returns the product of `area_e` and `dz_t`, which is the correct volume element for that grid cell location on a C grid. In my opinion we should change the test because, for example, this is how the vertical grid elements are handled in the MITgcm, so it seems like a legitimate case. 

Sorry this PR is dragging on! Let me know what you think!

Reply:
My intention was to first check if all `combined` metrics (e.g. area, cell_faces, volume) are correctly applied, but you are right: We should also test for datasets where certain combinations are not given as an array but instead as a combo of e.g. area and depth (the logic written by @rabernat, handles that very elegantly). As in my earlier comment, I think it would be best to expand this on top of #130, by dropping for instance the volume and then asserting that the returned metric is correctly reconstructed.